sycl/source/detail/program_manager/program_manager.cpp

```diff
@@ -110,22 +110,27 @@ DeviceImage &ProgramManager::getDeviceImage(OSModuleHandle M,
 }
 
 template <typename ExceptionT, typename RetT>
-RetT *
-waitUntilBuilt(KernelProgramCache &Cache,
-               KernelProgramCache::EntityWithState<RetT> *WithBuildState) {
+RetT *waitUntilBuilt(KernelProgramCache &Cache,
+                     KernelProgramCache::BuildResult<RetT> *BuildResult,
+                     bool &TryAgain) {
   // any thread which will find nullptr in cache will wait until the pointer
   // is not null anymore
-  Cache.waitUntilBuilt([WithBuildState]() {
-    int State = WithBuildState->State.load();
+  Cache.waitUntilBuilt([BuildResult]() {
+    int State = BuildResult->State.load();
 
     return State == BS_Done || State == BS_Failed;
   });
 
-  RetT *Result = WithBuildState->Ptr.load();
+  if (BuildResult->Error.FilledIn) {
+    const KernelProgramCache::BuildError &Error = BuildResult->Error;
+    throw ExceptionT(Error.Msg, Error.Code);
+  }
+
+  RetT *Result = BuildResult->Ptr.load();
 
-  if (!Result)
-    throw ExceptionT("The other thread tried to build the program/kernel but "
-                     "did not succeed.");
+  // if the result is still null then there was no SYCL exception and we may try
+  // to build kernel/program once more to generate the original exception
+  TryAgain = !Result;
```

Comment:
Do we need a separate parameter if `TryAgain` is always `!Result`?

Reply:
Yes, we don't really need it. Removed.